File: frontend/js/views/rent_escalation.js
// Lease cost with escalations + free rent — total, NPV, and effective monthly
// rent, via /calc/rent-escalation. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const VIEW = 'rent-escalation';
let lastReport = null;
let lastBody = null;

export async function renderRentEscalation(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rentesc.h1.title">// LEASE COST</span></h1>
        <p class="muted small" data-i18n="view.rentesc.hint.intro">
            The true cost of a lease with annual rent bumps and free-rent concessions. Rent steps up
            each anniversary; the first months can be waived. The effective monthly rent spreads the
            total over the full term — what you compare offers on. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.rentesc.h2.inputs">The lease</h2>
            <form id="rentesc-form" class="inline-form">
                <label><span data-i18n="view.rentesc.label.base">Base monthly rent ($)</span>
                    <input type="number" step="0.01" min="0" name="base_monthly_rent_usd" value="2000" required></label>
                <label><span data-i18n="view.rentesc.label.esc">Annual escalation (%)</span>
                    <input type="number" step="0.1" min="0" name="annual_escalation_pct" value="3" required></label>
                <label><span data-i18n="view.rentesc.label.term">Term (months)</span>
                    <input type="number" step="1" min="1" name="term_months" value="36" required></label>
                <label><span data-i18n="view.rentesc.label.free">Free months</span>
                    <input type="number" step="1" min="0" name="free_months" value="2"></label>
                <label><span data-i18n="view.rentesc.label.disc">Discount rate (%)</span>
                    <input type="number" step="0.01" min="0" name="discount_rate_pct" value="6"></label>
            </form>
            <div id="rentesc-tools" class="ce-toolbar"></div>
            <button type="button" id="rentesc-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="rentesc-sens" class="ce-sens"></div>
        </div>
        <div id="rentesc-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#rentesc-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            base_monthly_rent_usd: Number(fd.get('base_monthly_rent_usd')) || 0,
            annual_escalation_pct: Number(fd.get('annual_escalation_pct')) || 0,
            term_months: Number(fd.get('term_months')) || 0,
            free_months: Number(fd.get('free_months')) || 0,
            discount_rate_pct: Number(fd.get('discount_rate_pct')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcRentEscalation(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.rentesc.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#rentesc-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'rent-escalation.csv' });
    mount.querySelector('#rentesc-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['effective_monthly_rent_usd', r.effective_monthly_rent_usd],
        ['total_rent_usd', r.total_rent_usd],
        ['npv_usd', r.npv_usd],
        ['concession_value_usd', r.concession_value_usd],
        ['final_monthly_rent_usd', r.final_monthly_rent_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#rentesc-result');
    // Line chart: effective monthly rent as the annual escalation sweeps 0 -> 10%.
    const xs = enh.linspace(0, 10, 13);
    const pts = await Promise.all(xs.map(async (e) => {
        const rr = await api.calcRentEscalation({ ...body, annual_escalation_pct: e });
        return { x: e, y: rr ? rr.effective_monthly_rent_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'escalation %', ylabel: 'eff rent $/mo' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.rentesc.h2.result">The cost</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.rentesc.card.effective">Effective rent / mo</div>
                    <div class="value pos">${money(r.effective_monthly_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rentesc.card.total">Total rent</div>
                    <div class="value">${money(r.total_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rentesc.card.npv">NPV</div>
                    <div class="value">${money(r.npv_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.rentesc.row.total">Total rent</td><td>${money(r.total_rent_usd)}</td></tr>
                    <tr><td data-i18n="view.rentesc.row.npv">NPV of rent</td><td>${money(r.npv_usd)}</td></tr>
                    <tr><td data-i18n="view.rentesc.row.concession">Concession value</td><td>${money(r.concession_value_usd)}</td></tr>
                    <tr><td data-i18n="view.rentesc.row.final">Final month rent</td><td>${money(r.final_monthly_rent_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.rentesc.row.effective">Effective monthly rent</td><td>${money(r.effective_monthly_rent_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#rentesc-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: annual escalation 0 -> 10%; y: free months 0 -> 6. Output: effective monthly rent.
    const xVals = enh.linspace(0, 10, 5);
    const yVals = enh.linspace(0, 6, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_escalation_pct', yKey: 'free_months', xVals, yVals: yVals.map(Math.round), compute: (b) => api.calcRentEscalation(b), pick: (r) => (r ? r.effective_monthly_rent_usd : null) });
    if (!viewIsCurrent(tok)) return;
    // Lower effective rent is better (tenant), so negate for shading (green = cheaper).
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals: yVals.map(Math.round), cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : '$' + Math.round(-v)), xfmt: (v) => v.toFixed(0) + '%', yfmt: (v) => v.toFixed(0) + 'mo', xName: t('view.rentesc.label.esc') || 'Escal', yName: t('view.rentesc.label.free') || 'Free' });
}

File: crates/traderview-core/src/rent_escalation.rs
//! Lease cost with annual rent escalations and free-rent concessions — the
//! total and present value of a lease, and the effective (average) monthly
//! rent tenants compare offers on.
//!
//! Rent steps up at each anniversary: month `m` (1-based) sits in year
//! `(m−1)/12`, so `rent = base × (1 + escalation)^year`. The first
//! `free_months` are concessions (rent waived). The effective monthly rent
//! spreads the total over the full term, including the free months.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RentEscalationInput {
    pub base_monthly_rent_usd: f64,
    /// Annual rent escalation, percent (applied each anniversary).
    pub annual_escalation_pct: f64,
    pub term_months: u32,
    /// Free-rent months at the start (concession).
    #[serde(default)]
    pub free_months: u32,
    /// Annual discount rate for the present value, percent (0 = nominal only).
    #[serde(default)]
    pub discount_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RentEscalationResult {
    /// Total nominal rent paid over the term.
    pub total_rent_usd: f64,
    /// Present value of the rent stream at the discount rate.
    pub npv_usd: f64,
    /// Total ÷ term — the effective monthly rent.
    pub effective_monthly_rent_usd: f64,
    /// Value of the waived free-rent months (at the base rate).
    pub concession_value_usd: f64,
    /// Rent in the final month (fully escalated).
    pub final_monthly_rent_usd: f64,
}

pub fn analyze(input: &RentEscalationInput) -> RentEscalationResult {
    let esc = input.annual_escalation_pct / 100.0;
    let rm = input.discount_rate_pct / 100.0 / 12.0;

    let mut total = 0.0;
    let mut npv = 0.0;
    for m in 1..=input.term_months {
        let year = (m - 1) / 12;
        let mut rent = input.base_monthly_rent_usd * (1.0 + esc).powi(year as i32);
        if m <= input.free_months {
            rent = 0.0;
        }
        total += rent;
        npv += rent / (1.0 + rm).powi(m as i32);
    }

    let effective = if input.term_months > 0 {
        total / input.term_months as f64
    } else {
        0.0
    };
    let final_rent = if input.term_months > 0 {
        let year = (input.term_months - 1) / 12;
        input.base_monthly_rent_usd * (1.0 + esc).powi(year as i32)
    } else {
        0.0
    };

    RentEscalationResult {
        total_rent_usd: total,
        npv_usd: npv,
        effective_monthly_rent_usd: effective,
        concession_value_usd: input.free_months as f64 * input.base_monthly_rent_usd,
        final_monthly_rent_usd: final_rent,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> RentEscalationInput {
        RentEscalationInput {
            base_monthly_rent_usd: 2000.0,
            annual_escalation_pct: 3.0,
            term_months: 36,
            free_months: 2,
            discount_rate_pct: 6.0,
        }
    }

    #[test]
    fn total_rent() {
        assert!(close(analyze(&base()).total_rent_usd, 70181.60));
    }

    #[test]
    fn effective_monthly() {
        assert!(close(analyze(&base()).effective_monthly_rent_usd, 1949.4889));
    }

    #[test]
    fn npv_below_nominal() {
        let r = analyze(&base());
        assert!(close(r.npv_usd, 63684.0024));
        assert!(r.npv_usd < r.total_rent_usd);
    }

    #[test]
    fn final_month_fully_escalated() {
        // Year 2 (months 25-36): 2000 × 1.03² = 2121.80.
        assert!(close(analyze(&base()).final_monthly_rent_usd, 2121.80));
    }

    #[test]
    fn concession_value() {
        assert!(close(analyze(&base()).concession_value_usd, 4000.0));
    }

    #[test]
    fn no_escalation_is_flat() {
        let r = analyze(&RentEscalationInput {
            annual_escalation_pct: 0.0,
            free_months: 0,
            discount_rate_pct: 0.0,
            ..base()
        });
        // 36 × 2000 = 72,000; effective equals base.
        assert!(close(r.total_rent_usd, 72_000.0));
        assert!(close(r.effective_monthly_rent_usd, 2000.0));
    }

    #[test]
    fn free_months_lower_effective_rent() {
        let with_free = analyze(&base());
        let no_free = analyze(&RentEscalationInput {
            free_months: 0,
            ..base()
        });
        assert!(with_free.effective_monthly_rent_usd < no_free.effective_monthly_rent_usd);
    }

    #[test]
    fn zero_discount_npv_equals_total() {
        let r = analyze(&RentEscalationInput {
            discount_rate_pct: 0.0,
            ..base()
        });
        assert!(close(r.npv_usd, r.total_rent_usd));
    }
}
